modules/parachains/src/lib.rs

```diff
@@ -0,0 +1,594 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Parachains finality module.
+//!
+//! This module needs to be deployed with GRANDPA module, which is syncing relay
+//! chain blocks. The main entry point of this module is `submit_parachain_heads`, which
+//! accepts storage proof of some parachain `Heads` entries from bridged relay chain.
+//! It requires corresponding relay headers to be already synced.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use bp_polkadot_core::parachains::{
+	parachain_head_hash, ParaHash, ParaHead, ParaId, ParachainHeadsProof,
+};
+use codec::{Decode, Encode};
+use frame_support::RuntimeDebug;
+use scale_info::TypeInfo;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::vec::Vec;
+
+// Re-export in crate namespace for `construct_runtime!`.
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+/// Block hash of the bridged relay chain.
+pub type RelayBlockHash = bp_polkadot_core::Hash;
+/// Block number of the bridged relay chain.
+pub type RelayBlockNumber = bp_polkadot_core::BlockNumber;
+/// Hasher of the bridged relay chain.
+pub type RelayBlockHasher = bp_polkadot_core::Hasher;
+
+/// Best known parachain head as it is stored in the runtime storage.
+#[derive(Decode, Encode, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct BestParaHead {
+	/// Number of relay block where this head has been updated.
+	pub at_relay_block_number: RelayBlockNumber,
+	/// Hash of parachain head.
+	pub head_hash: ParaHash,
+	/// Current ring buffer position for this parachain.
+	pub next_imported_hash_position: u32,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// Relay chain block is unknown to us.
+		UnknownRelayChainBlock,
+		/// Invalid storage proof has been passed.
+		InvalidStorageProof,
+	}
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config<I: 'static = ()>:
+		pallet_bridge_grandpa::Config<Self::BridgesGrandpaPalletInstance>
+	{
+		/// Instance of bridges GRANDPA pallet that this pallet is linked to.
+		///
+		/// The GRANDPA pallet instance must be configured to import headers of relay chain that
+		/// we're interested in.
+		type BridgesGrandpaPalletInstance: 'static;
+
+		/// Maximal number of single parachain heads to keep in the storage.
+		///
+		/// The setting is there to prevent growing the on-chain state indefinitely. Note
+		/// the setting does not relate to parachain block numbers - we will simply keep as much
+		/// items in the storage, so it doesn't guarantee any fixed timeframe for heads.
+		#[pallet::constant]
+		type HeadsToKeep: Get<u32>;
+	}
+
+	/// Best parachain heads.
+	#[pallet::storage]
+	pub type BestParaHeads<T: Config<I>, I: 'static = ()> =
```

Comment:
Identity hasher for `ParaId` should probably be replaced with blake2 here and below

Reply:
I've changed all to be the strongest `Blake2_128Concat` for `ParaId` and `ParaHash`, as they can be (in theory) selected by the submitter. As for the insertion index, I've changed to `Twox64Concat`